Core/include/Acts/TrackFitting/Chi2Fitter.hpp

```diff
@@ -0,0 +1,748 @@
+        // Evaluate the material properties
+        if (interaction.evaluateMaterialSlab(state, updateStage)) {
+          // Surface has material at this stage
+          hasMaterial = true;
+
+          // Evaluate the material effects
+          interaction.evaluatePointwiseMaterialInteraction(multipleScattering,
+                                                           energyLoss);
+
+          // Screen out material effects info
+          ACTS_VERBOSE("Material effects on surface: "
+                       << surface->geometryId()
+                       << " at update stage: " << updateStage << " are :");
+          ACTS_VERBOSE("eLoss = "
+                       << interaction.Eloss << ", "
+                       << "variancePhi = " << interaction.variancePhi << ", "
+                       << "varianceTheta = " << interaction.varianceTheta
+                       << ", "
+                       << "varianceQoverP = " << interaction.varianceQoverP);
+
+          // Update the state and stepper with material effects
+          interaction.updateState(state, stepper);
+        }
+      }
+
+      if (not hasMaterial) {
+        // Screen out message
+        ACTS_VERBOSE("No material effects on surface: " << surface->geometryId()
+                                                        << " at update stage: "
+                                                        << updateStage);
+      }
+    }
+  };
+
+  template <typename parameters_t>
+  class Aborter {
+   public:
+    /// Broadcast the action_type
+    using action_type = Actor<parameters_t>;
+
+    template <typename propagator_state_t, typename stepper_t,
+              typename result_t>
+    bool operator()(propagator_state_t& /*state*/, const stepper_t& /*stepper*/,
+                    const result_t& result) const {
+      // const auto& logger = state.options.logger;
+      if (!result.result.ok() or result.finished) {
+        return true;
+      }
+      return false;
+    }
+  };
+
+ public:
+  /// Fit implementation of the GX2F
+  ///
+  /// @tparam source_link_iterator_t Iterator type used to pass source links
+  /// @tparam start_parameters_t Type of the initial parameters
+  /// @tparam parameters_t Type of parameters used for local parameters
+  ///
+  /// @param it Begin iterator for the fittable uncalibrated measurements
+  /// @param end End iterator for the fittable uncalibrated measurements
+  /// @param sParameters The initial track parameters
+  /// @param chi2FitterOptions Chi2FitterOptions steering the fit
+  ///
+  /// @note The input measurements are given in the form of @c SourceLink s.
+  /// It's the calibrators job to turn them into calibrated measurements used in
+  /// the fit.
+  ///
+  /// @return the output as an output track
+  template <typename source_link_iterator_t, typename start_parameters_t,
+            typename parameters_t = BoundTrackParameters>
+  Result<Chi2FitterResult<traj_t>> fit(
```

Comment:
Following the style of the other fitters (KF and GSF), this should maybe also take a `std::shared_ptr<traj_t>` as an (optional) argument?